src/moneypenny/kubernetes.py

```diff
@@ -1,165 +1,315 @@
 """Kubernetes client abstraction for Moneypenny."""
-
+import json
 from pathlib import Path
 from typing import Any, Dict, List, Optional
 
 import structlog
-from yaml import safe_load
 from kubernetes.client import (
     V1Capabilities,
+    V1ConfigMap,
+    V1ConfigMapVolumeSource,
     V1Container,
     V1LocalObjectReference,
     V1ObjectMeta,
+    V1Pod,
     V1PodSecurityContext,
     V1PodSpec,
     V1SecurityContext,
+    V1Volume,
 )
 from kubernetes.client.api import core_v1_api
 from kubernetes.client.exceptions import ApiException
+from kubernetes.config import load_incluster_config, load_kube_config
+from kubernetes.config.config_exception import ConfigException
 
-from .types import OperationFailed, PodNotFound, K8sApiException
-
-NAMESPACE_FILE = "/var/run/secrets/kubernetes.io/serviceaccount/namespace"
-PRIMARY_CONTAINER_IMAGE = "library/alpine:latest"
+from .config import Configuration
+from .errors import K8sApiException, OperationFailed, PodNotFound
 
 logger = structlog.get_logger(__name__)
 
 
+def _name_object(username: str, type: str) -> str:
+    """This constructs a consistent object name from the username and
+    type.  Purely syntactic sugar, but tasty and widely used in here.
+    """
+    return f"{username}-{type}"
+
+
 class KubernetesClient:
     """A client for moneypenny's usage of Kubernetes.
 
     This provides a level of abstraction away from the python-kubernetes
     objects and models.  By hiding these obscure objects here it makes
     the code easier to mock and test.
+
+    The public API is just three methods: make_objects, delete_objects,
+    and check_pod_completed.
+
+    All Exceptions raised are our own, although they may well be thin
+    wrappers around the corresponding K8s API error.
     """
 
-    def __init__(self) -> None:
-        """Create a new client for the cluster we are running in."""
-        self.namespace = Path(NAMESPACE_FILE).read_text().strip()
+    def __init__(self, moneypenny) -> None:  # type: ignore
+        """Create a new client for the cluster we are running in.
+        We ignore type in order to prevent a circular dependency."""
```

Comment:
You're only passing in the moneypenny object to get the config. Just pass in the config directly and that avoids the circular dependency. The Kubernetes client should never need to know about the moneypenny object; that would be a layering violation.

Reply:
Yes completely agree with this.  I don't even think you need to pass in the config I think you could make a new Configuration class here and just load it right from the environment variables.  Sure it loads the environment variables a couple of times, but they aren't going to change.

Reply:
I'm passing the config down.  I guess I *could* make one there, but...meh.  Six of one...